README.md
# rasactl

`rasactl` deploys Rasa X / Enterprise on your local or remote Kubernetes cluster and manages Rasa X / Enterprise deployments.

<p align="center"><img src="img/render1631526529716.gif?raw=true"/></p>

## Features

- deploy Rasa X / Enterprise

  You can use `rasactl` to deploy Rasa X / Enterprise on your local machine or a VM in one of the major cloud providers.

  (check the [Prerequisites](#Prerequisites) section)

- upgrade Rasa X / Enterprise

  Upgrade/change configuration for an existing Rasa X deployment.

- stop/delete/start a running Rasa X / Enterprise deployment

  Manage the lifecycle of your deployment: you can stop, delete or start one of the Rasa X deployments managed by `rasactl`.

- connect a local Rasa Server to Rasa X / Enterprise

  You can use your local Rasa Open Source server along with Rasa X / Enterprise. `rasactl` will prepare configuration for Rasa OSS and Rasa X and run the Rasa Open Source server on your local machine.

  (requires `kind` and Rasa OSS installed locally)

- use a local Rasa project along Rasa X / Enterprise deployment

  Use your local Rasa project along with Rasa X / Enterprise deployment. The `rasactl` provides an easy way to use your local Rasa project along with Rasa X / Enterprise.

  This setup was previously referred to as "local mode" in older Rasa X versions.

  (requires `kind` and Rasa X >= 1.0.0)

## Table of Contents

- [rasactl](#rasactl)
  - [Features](#features)
  - [Table of Contents](#table-of-contents)
  - [Prerequisites](#prerequisites)
  - [Installation](#installation)
    - [Linux / macOS](#linux--macos)
  - [Compatibility matrix](#compatibility-matrix)
  - [Before you start](#before-you-start)
  - [Values File](#values-file)
  - [Configuration](#configuration)
    - [Environment variables](#environment-variables)
    - [Configuration file](#configuration-file)
  - [Global flags](#global-flags)
  - [Commands](#commands)
    - [The `add` command](#the-add-command)
    - [The `start` command](#the-start-command)
    - [The `stop` command](#the-stop-command)
    - [The `delete` command](#the-delete-command)
    - [The `list` command](#the-list-command)
    - [The `status` command](#the-status-command)
    - [The `config use-deployment` command](#the-config-use-deployment-command)
    - [The `connect rasa` command](#the-connect-rasa-command)
    - [The `auth login` command](#the-auth-login-command)
    - [The `auth logout` command](#the-auth-logout-command)
    - [The `logs` command](#the-logs-command)
  - [Enterprise Management Commands](#enterprise-management-commands)
    - [The `enterprise activate` command](#the-enterprise-activate-command)
    - [The `enterprise deactivate` command](#the-enterprise-deactivate-command)
  - [Model Management Commands](#model-management-commands)
    - [The `model delete` command](#the-model-delete-command)
    - [The `model download` command](#the-model-download-command)
    - [The `model list` command](#the-model-list-command)
    - [The `model tag` command](#the-model-tag-command)
    - [The `model upload` command](#the-model-upload-command)
    - [Upload a model to Rasa X](#upload-a-model-to-rasa-x)
  - [Examples of usage](#examples-of-usage)
    - [Run Rasa X / Enterprise with a local Rasa Server](#run-rasa-x--enterprise-with-a-local-rasa-server)
    - [Run Rasa X / Enterprise with mounted a local Rasa project](#run-rasa-x--enterprise-with-mounted-a-local-rasa-project)
    - [Upgrade Rasa X / Enterprise version](#upgrade-rasa-x--enterprise-version)
    - [Deploy Rasa X in one of the public cloud providers](#deploy-rasa-x-in-one-of-the-public-cloud-providers)
  - [Development](#development)
    - [How to run it?](#how-to-run-it)
    - [Run unit tests](#run-unit-tests)
    - [Kind cluster for developing purposes](#kind-cluster-for-developing-purposes)
  - [License](#license)

## Prerequisites

- Kubernetes cluster >= 1.14+

   or

- kind (for local mode)

(You can use the [REI](https://github.com/RasaHQ/REI) to install all required components on your local machine or a VM.)

## Installation

### Linux / macOS

- Binary downloads of `rasactl` can be found on [the Releases page](https://github.com/rasahq/rasactl/releases/latest). You can manually install rasactl by coping the binary into your `bin`:

```text
$ curl -L https://github.com/RasaHQ/rasactl/releases/download/0.0.24/rasactl_0.0.24_darwin_amd64.tar.gz -O
$ tar -zxvf rasactl_0.0.24_darwin_amd64.tar.gz
$ cp rasactl_0.0.24_darwin_amd64/rasactl /usr/local/bin/
```

- You can also install via `brew`:

```text
$ brew tap rasahq/rasactl
$ brew install rasactl
```

## Compatibility matrix

| rasactl version | Helm chart version |
| --------------- | ------------------ |
| `>= 1.0.x`      | `4.x`              |
| `<= 0.5.x`      | `3.x`              |

## Before you start

Below you can find several things that are good to know and keep in mind when you use `rasactl`.

- It is possible to configure multiple deployments with `rasactl`. A `rasactl` command will always execute an operation on a single deployment. Here is the order in which `rasactl` determines which deployment to use:

  1. A deployment name passed as an argument in CLI, e.g. `rasactl status deployment-name`, you can use `rasactl help command` to see usage example for a given command.
  2. `rasactl` checks if a `.rasactl` file exists in a current working directory. If so, the deployment defined in the file is used. This `.rasactl` file is created automatically when the `rasactl start --project` command is executed.
  3. `rasactl` checks if a default deployment is configured in the `rasactl.yaml` configuration file, if yes, then the default deployment is used. The default deployment can be set by using the `rasactl config use-deployment` command.
  4. If there is only one deployment, then it's used.

  You can use the [`rasactl list`](#the-list-command) command to check which deployment is used as the current one.

  The `rasactl delete` command requires explicitly passing a deployment name as an argument.

- `rasactl` uses the [`rasa-x-helm`](https://github.com/RasaHQ/rasa-x-helm) chart to deploy Rasa X / Enterprise.
- `rasactl` deploys Rasa X / Enterprise without a Rasa Open Source server. It's up to you to connect Rasa OSS with Rasa X / Enterprise deployment.
- `rasactl` uses a Kubernetes context from the kubeconfig file, if you want to switch Kubernetes cluster you have to use `kubectl` or other tools that change the active context for the kubeconfig.

## Values File

The `rasactl` uses the [`rasa-x-helm` chart](https://github.com/RasaHQ/rasa-x-helm) to deploy Rasa X / Enterprise, which means you can use [the helm chart values](https://github.com/RasaHQ/rasa-x-helm/blob/main/charts/rasa-x/values.yaml) to configure deployment. The `rasactl` enables template usage for the values file so that it's possible to use the [Go template](https://pkg.go.dev/text/template#hdr-Actions) and [Sprig function](http://masterminds.github.io/sprig/) within the value file, e.g.

```yaml
# values.yaml
rasax:
  podLabels:
    rasactl: "true"
    test_version: {{ env "RASACTL_TEST_VERSION" }}
    test_template: {{ coalesce 0 1 2 }}
```

## Configuration

### Environment variables

|                  Name                  |                                                                                                                         Description                                                                                                                          |
| -------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `RASACTL_AUTH_USER`                    | The username that is used to authorize to Rasa X / Enterprise                                                                                                                                                                                                |
| `RASACTL_AUTH_PASSWORD`                | The password that is used to authorize to Rasa X / Enterprise                                                                                                                                                                                                |
| `RASACTL_RASA_X_URL`                   | Set Rasa X / Enterprise URL. By default, the URL is detected automatically, but if you use a custom configuration and you wanna define Rasa X URL explicitly you can use the env variable. The `RASACTL_RASA_X_URL` overrides Rasa X URL for all deployment. |
| `RASACTL_RASA_X_URL_<DEPLOYMENT_NAME>` | Set Rasa X / Enterprise URL for a given deployment, e.g. if a deployment name is `my-deployment`, then you can use the `RASACTL_RASA_X_URL_MY_DEPLOYMENT` environment variable to define the Rasa X URL for the `my-deployment`.                             |
| `RASACTL_KUBECONFIG`                   | Absolute path to the kubeconfig file (default "`$HOME/.kube/config`")                                                                                                                                                                                        |

### Configuration file

Below you can find an example of the configuration file and parameters that can be defined, by default configuration file is located in `$HOME/.rasactl.yaml`.

```yaml
# Deployment name that is used as a current deployment (default).
# You can use the `rasactl config use-deployment` command to set the current deployment.
current-deployment: my-deployment

# Name of the kubeconfig context to use
kube-context: ""

# Absolute path to the kubeconfig file
kubeconfig: /home/user/.kube/config
```

## Global flags

Below you can find global flags that can be used with every command.

```text
Global Flags:
      --config string         config file (default is $HOME/.rasactl.yaml)
      --debug                 enable debug output
  -h, --help                  help for rasactl
      --kube-context string   name of the kubeconfig context to use
      --kubeconfig string     absolute path to the kubeconfig file (default "$HOME/.kube/config")
      --verbose               enable verbose output
```

## Commands

```text
Available Commands:
  add         add existing Rasa X deployment to rasactl
  auth        manage credentials for Rasa X / Enterprise
  completion  generate the autocompletion script for the specified shell
  config      modify the configuration file
  connect     connect a component (e.g. a Rasa OSS server) to Rasa X
  delete      delete Rasa X deployment
  enterprise  manage Rasa Enterprise
  help        Help about any command
  list        list deployments
  logs        print the logs for a container in a pod
  model       manage models for Rasa X / Enterprise
  open        open Rasa X in a web browser
  start       start a Rasa X deployment
  status      show deployment status
  stop        stop Rasa X deployment
  upgrade     upgrade Rasa X deployment
```

### The `add` command

Adds existing Rasa X deployment to rasactl.

If you already have a Rasa X deployment that uses the rasa-x-helm chart you can add the deployment and manage it by rasactl.

```text
Usage:
  rasactl add NAMESPACE [flags]
```

```text
Examples:
  # Add a Rasa X deployment that is deployed in the 'my-test' namespace.
  $ rasactl add my-test

  # Add a Rasa X deployment that is deployed in the 'my-test' namespace and
  # a helm release name for the deployment is 'rasa-x-example'.
  $ rasactl add my-test --rasa-x-release-name rasa-x-example
```

```text
Flags:
  -h, --help                         help for add
      --rasa-x-release-name string   a helm release name to manage (default "rasa-x")
```

### The `start` command

The `start` command creates a Rasa X deployment or starts a stopped deployment if a given deployment already exists.

```text
Usage:
  rasactl start [DEPLOYMENT-NAME] [flags]

```

```text
Examples:
  # Create a new Rasa X deployment with an autogenerated name.
  $ rasactl start

  # Create a Rasa X deployment with a defined name.
  $ rasactl start my-deployment

  # Create a new deployment if there is already one or more deployments.
  # rasactl start --create

  # Create a Rasa X deployment with custom configuration, e.g the following configuration changes a Rasa X version.
  # All available values: https://github.com/RasaHQ/rasa-x-helm/blob/main/charts/rasa-x/values.yaml
  $ rasactl start --values-file custom-configuration.yaml

  # Create a Rasa X deployment with a defined password.
  $ rasactl start --rasa-x-password mypassword

  # Create a Rasa X deployment that uses a local Rasa project.
  # The command is executed in a Rasa project directory.
  $ rasactl start --project
```

```text
Flags:
      --create                        create a new deployment. If --project or --project-path is set, or there is no existing deployment, the flag is not required to create a new deployment
  -h, --help                          help for start
  -p, --project                       use the current working directory as a project directory, the flag is ignored if --project-path is used
      --project-path string           absolute path to the project directory mounted in kind
      --rasa-x-chart-version string   a helm chart version to use
      --rasa-x-edge-release           use the latest edge release of Rasa X
      --rasa-x-password string        Rasa X password (default "rasaxlocal")
      --rasa-x-password-stdin         read the Rasa X password from stdin
      --rasa-x-release-name string    a helm release name to manage (default "rasa-x")
      --values-file string            absolute path to the values file
      --wait-timeout duration         time to wait for Rasa X to be ready (default 10m0s)
```

### The `stop` command

The `stop` command stops a running Rasa X / Enterprise deployment. The Rasa X deployment and all its components will be scaled down to 0.

```text
Usage:
  rasactl stop [DEPLOYMENT-NAME] [flags]
```

```text
Examples:
  # Stop a Rasa X deployment with the 'my-deployment' name.
  $ rasactl stop my-deployment

  # Stop a currently active Rasa X deployment.
  # The command stops the currently active deployment.
  # You can use the 'rasactl list' command to check which deployment is currently used.
  $ rasactl stop
```

```text
Flags:
  -h, --help   help for stop
```

### The `delete` command

The `delete` command deletes a Rasa X / Enterprise deployment.

You can use the `--prune` flag to remove a namespace where Rasa X deployment is located.

***Notice*** If you want to free resources, or temporarily you don't need to run Rasa X deployment, you can stop a Rasa X / Enterprise deployment instead of deleting it. Stopping the Rasa X deployment will free resources, but keep the current configuration.

```text
Usage:
  rasactl delete DEPLOYMENT-NAME [flags]

Aliases:
  delete, del
```

```text
Examples:
  # Delete the 'my-example' deployment.
  $ rasactl delete my-example

  # Prune the 'my-example' deployment, execute the command with the --prune flag deletes the whole namespace.
  $ rasactl delete my-example --prune
```

```text
Flags:
      --force   if true, delete resources and ignore errors
  -h, --help    help for delete
      --prune   if true, delete a namespace with a project
```

### The `list` command

List all deployments.

```text
$ rasactl list
CURRENT	NAME         	STATUS 	RASA PRODUCTION	RASA WORKER	ENTERPRISE	VERSION
       	hopeful-haibt	Running	2.8.1          	2.8.1      	inactive  	0.42.0
*      	vibrant-yalow	Running	2.8.1          	2.8.1      	inactive  	0.42.0
```

The `*` in the `CURRENT` field indicates a deployment that is used as default. It means that every time when you execute `rasactl` command without defining the deployment name, the deployment marked with `*` is used.

A deployment is marked as `CURRENT` if:

- there is a `.rasactl` file that includes a deployment name in your current working directory. The file is automatically created if you run the `rasactl start` command with the `--project` or `--project-path` flag
- there is only one deployment
- you set the current deployment by using the `rasactl config use-deployment` command

### The `status` command

Show the status of a deployment.

```text
Usage:
  rasactl status [DEPLOYMENT-NAME] [flags]
```

```text
Examples:
  # Show status for the 'example' deployment.
  $ rasactl status example

  # Show status for the 'example' deployment along with details.
  $ rasactl status example --details
```

```text
Flags:
  -d, --details         show detailed information, such as running pods, helm chart status
  -h, --help            help for status
  -o, --output string   output format. One of: json|table (default "table")
```

Example output:

```text
$ rasactl status vibrant-yalow
Name:                   	vibrant-yalow
Status:                 	Running
URL:                    	http://vibrant-yalow.rasactl.localhost
Version:                	0.42.0
Enterprise:             	inactive
Rasa production version:	2.8.1
Rasa worker version:    	2.8.1
Project path:           	/home/ubuntu/test
```

### The `config use-deployment` command

Sets the current-deployment in the configuration file.

If you have multiple Rasa X deployments, and you are not in a project directory you have to explicitly define the deployment name during command execution.
You can define a deployment that is used as a current one by using the `rasa config use-deployment` command.

```text
Usage:
  rasactl config use-deployment DEPLOYMENT-NAME [flags]
```

```text
Examples:
  # Set the 'example' deployment as the current deployment.
  $ rasactl config use-deployment example
```

```text
Flags:
  -h, --help   help for use-deployment
```

### The `connect rasa` command

Run a local Rasa Open Source server and connect it to a Rasa X deployment.

The command prepares a configuration that's required to connect Rasa X deployment and run a local Rasa server.

It's required to have the 'rasa' command accessible by rasactl.

The command works only if Rasa X deployment runs on a local Kubernetes cluster managed with 'kind'.

```text
Usage:
  rasactl connect rasa [DEPLOYMENT-NAME] [flags]
```

```text
Examples:
  # Connect Rasa Server to Rasa X deployment.
  $ rasactl connect rasa

  # Run a separate rasa server for the Rasa X worker environment.
  $ rasactl connect rasa --run-separate-worker

  # Pass extra arguments to rasa server.
  $ rasactl connect rasa --extra-args="--debug"
```

```text
Flags:
      --extra-args strings    extra arguments for Rasa server
  -h, --help                  help for rasa
  -p, --port int              port to run the Rasa server at (default 5005)
      --run-separate-worker   runs a separate Rasa server for the worker environment
```

### The `auth login` command

Log in to Rasa X / Enterprise.

`auth login` stores credentials in an external credentials store, such as the native keychain of the operating system.

 The following external credential stores will be used:

  *  On macOS: [Apple macOS Keychain Access](https://support.apple.com/en-gb/guide/keychain-access/welcome/mac)
  *  On Linux: [pass](https://www.passwordstore.org/)
  *  On Windows: [Microsoft Windows Credential Manager](https://support.microsoft.com/en-us/windows/accessing-credential-manager-1b5c916a-6a16-889f-8581-fc16e8165ac0)

 You can pass credentials via environment variables:

  *  `RASACTL_AUTH_USER` - username
  *  `RASACTL_AUTH_PASSWORD` - password

 If the environment variables are used, credentials stored in a native keychain are not used.

```text
Usage:
  rasactl auth login [DEPLOYMENT-NAME] [flags]
```

```text
Examples:
  # Login to the 'my-deployment' Rasa X / Enterprise deployment.
  $ rasactl auth login my-deployment

  # Login to Rasa X / Enterprise (login to the currently active deployment).
  $ rasactl auth login

  # Provide a password using STDIN.
  # You can login non-interactively by using the --password-stdin flag to provide a password through STDIN.
  # Using STDIN prevents the password from ending up in the shell’s history.
  $ rasactl auth login --username me --password-stdin
```

```text
Flags:
  -h, --help              help for login
  -p, --password string   password
      --password-stdin    read the password from stdin
  -u, --username string   username
```

***Notice*** For Linux, `pass` is used as credential storage. `pass` must be installed and configured before you use the `rasactl auth` command. Below you can find an example of `pass` installation and configuration.

`pass` installation and configuration for Linux Ubuntu.

1. Install `pass`.

```text
sudo apt-get install pass
```

2. Generate a GPG key.

```text
$  gpg --gen-key
gpg (GnuPG) 2.2.19; Copyright (C) 2019 Free Software Foundation, Inc.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

gpg: directory '/home/ubuntu/.gnupg' created
gpg: keybox '/home/ubuntu/.gnupg/pubring.kbx' created
Note: Use "gpg --full-generate-key" for a full featured key generation dialog.

GnuPG needs to construct a user ID to identify your key.

Real name: rasactl
Email address:
You selected this USER-ID:
    "rasactl"

Change (N)ame, (E)mail, or (O)kay/(Q)uit? O
[...]
public and secret key created and signed.
```

3. Init `pass`.

```text
$ pass init rasactl
mkdir: created directory '/home/ubuntu/.password-store/'
Password store initialized for rasactl
```

Now you can use `rasactl auth` on Linux.

```text
$ rasactl ls
CURRENT	NAME             	STATUS 	RASA PRODUCTION	RASA WORKER	ENTERPRISE	VERSION
*       wonderful-gagarin	Running	2.8.1          	2.8.1      	inactive  	0.42.0
$ rasactl auth login
Username: me
Password:
Successfully logged.
```

***Troubleshooting*** If you see `Error: exit status 2: gpg: decryption failed: No secret key` error you should export the following environment variable `export GPG_TTY="$(tty)"`.

### The `auth logout` command

Removes credentials from an external credentials store, such as the native keychain of the operating system.

```text
Usage:
  rasactl auth logout [DEPLOYMENT-NAME] [flags]
```

```text
Examples:
  # Remove access credentials (use the currently active deployment).
  $ rasactl auth logout

  # Remove access credentials for the 'my-deployment' deployment.
  $ rasactl auth logout my-deployment
```

```text
Flags:
  -h, --help   help for logout
```

### The `logs` command

Print the logs for a container in a pod. If the pod has only one container, the container name is optional.

```text
Usage:
  rasactl logs [DEPLOYMENT-NAME] [POD] [flags]
```

```text
Examples:
  # Choose a pod and show logs for it (use the currently active deployment).
  $ rasactl logs

  # Show logs from pod rasa-x (use the currently active deployment).
  $ rasactl logs rasa-x

  # Show logs from pod rasa-x for the 'my-deployment' deployment.
  $ rasactl logs my-deployment rasa-x

  # Display only the most recent 10 lines of output in pod rasa-x
  $ rasactl logs rasa-x --tail=10

  # Return snapshot of previous terminated nginx container logs from pod rasa
  $ rasactl logs -p -c nginx rasa

  # Begin streaming the logs from pod rasa-x
  $ rasactl logs -f rasa-x
```

```text
Flags:
  -c, --container string   a container name
  -f, --follow             specify if the logs should be streamed
  -h, --help               help for logs
  -p, --previous           print the logs for the previous instance of the container in a pod if it exists
      --tail int           lines of recent log file to display. Defaults to -1 showing all log lines (default -1)
```

## Enterprise Management Commands

You can manage an Enterprise license via `rasactl`.

```text
manage Rasa Enterprise

Usage:
  rasactl enterprise [command]

Available Commands:
  activate    activate an Enterprise license
  deactivate  deactivate an Enterprise license
```

### The `enterprise activate` command

Activate an Enterprise license.

```text
Usage:

CHANGELOG.md

<a name="1.0.4"></a>
## [1.0.4](https://github.com/RasaHQ/rasactl/compare/1.0.3...1.0.4) (2022-02-07)

### Bug Fixes

* Return valid deployment status if helm release is different than the default ([#40](https://github.com/RasaHQ/rasactl/issues/40))
* Set default helm chart version to 4.3.3 ([#41](https://github.com/RasaHQ/rasactl/issues/41))

### Code Refactoring

* Use xerrors package ([#38](https://github.com/RasaHQ/rasactl/issues/38))


<a name="1.0.3"></a>
## [1.0.3](https://github.com/RasaHQ/rasactl/compare/1.0.2...1.0.3) (2022-01-27)

### Bug Fixes

* Support for using different helm release name and versions ([#31](https://github.com/RasaHQ/rasactl/issues/31))
* Return a URL if a web browser is not detected ([#29](https://github.com/RasaHQ/rasactl/issues/29))


<a name="1.0.2"></a>
## [1.0.2](https://github.com/RasaHQ/rasactl/compare/1.0.1...1.0.2) (2022-01-25)

### Bug Fixes

* Use hostAliases for rasa-x pod instead of extraHosts for docker ([#27](https://github.com/RasaHQ/rasactl/issues/27))
* Use local RASA X URL if it connects to local Rasa OSS ([#26](https://github.com/RasaHQ/rasactl/issues/26))


<a name="1.0.1"></a>
## [1.0.1](https://github.com/RasaHQ/rasactl/compare/1.0.0...1.0.1) (2022-01-22)

### Bug Fixes

* Use correct Helm value path for rabbitmq credentials ([#23](https://github.com/RasaHQ/rasactl/issues/23))


<a name="1.0.0"></a>
## [1.0.0](https://github.com/RasaHQ/rasactl/compare/0.5.0...1.0.0) (2022-01-19)

### Breaking changes

* Update default helm values ([#17](https://github.com/RasaHQ/rasactl/issues/17))

### Bug Fixes

* Update dependencies ([#18](https://github.com/RasaHQ/rasactl/issues/18))


<a name="0.5.0"></a>
## [0.5.0](https://github.com/RasaHQ/rasactl/compare/0.4.3...0.5.0) (2022-01-05)

### Features

* Add additional deployment statuses (upgrading & installing) ([#16](https://github.com/RasaHQ/rasactl/issues/16))


<a name="0.4.3"></a>
## [0.4.3](https://github.com/RasaHQ/rasactl/compare/0.4.2...0.4.3) (2021-12-02)

### Bug Fixes

* Check if a namespace is managed by rasactl on delete ([#15](https://github.com/RasaHQ/rasactl/issues/15))


<a name="0.4.2"></a>
## [0.4.2](https://github.com/RasaHQ/rasactl/compare/0.4.1...0.4.2) (2021-11-30)


<a name="0.4.1"></a>
## [0.4.1](https://github.com/RasaHQ/rasactl/compare/0.4.0...0.4.1) (2021-11-18)

### Bug Fixes

* Fix adding a label to namespace if the labels field is absent ([#11](https://github.com/RasaHQ/rasactl/issues/11))


<a name="0.4.0"></a>
## [0.4.0](https://github.com/RasaHQ/rasactl/compare/0.3.0...0.4.0) (2021-10-18)

### Features

* Enable template for the values file ([#9](https://github.com/RasaHQ/rasactl/issues/9))
* Set Rasa X URL via environment variables  ([#8](https://github.com/RasaHQ/rasactl/issues/8))


<a name="0.3.0"></a>
## [0.3.0](https://github.com/RasaHQ/rasactl/compare/0.2.0...0.3.0) (2021-10-15)

### Features

* Add an option to pass creds via env variables ([#7](https://github.com/RasaHQ/rasactl/issues/7))


<a name="0.2.0"></a>
## [0.2.0](https://github.com/RasaHQ/rasactl/compare/0.2.0-rc.2...0.2.0) (2021-10-14)


<a name="0.2.0-rc.2"></a>
## [0.2.0-rc.2](https://github.com/RasaHQ/rasactl/compare/0.2.0-rc.1...0.2.0-rc.2) (2021-10-12)


<a name="0.2.0-rc.1"></a>
## [0.2.0-rc.1](https://github.com/RasaHQ/rasactl/compare/0.1.0...0.2.0-rc.1) (2021-10-07)


<a name="0.1.0"></a>
## [0.1.0](https://github.com/RasaHQ/rasactl/compare/0.0.26...0.1.0) (2021-09-23)


<a name="0.0.26"></a>
## [0.0.26](https://github.com/RasaHQ/rasactl/compare/0.0.25...0.0.26) (2021-09-22)


<a name="0.0.25"></a>
## [0.0.25](https://github.com/RasaHQ/rasactl/compare/0.0.24...0.0.25) (2021-09-21)


<a name="0.0.24"></a>
## [0.0.24](https://github.com/RasaHQ/rasactl/compare/0.0.23...0.0.24) (2021-09-20)


<a name="0.0.23"></a>
## [0.0.23](https://github.com/RasaHQ/rasactl/compare/0.0.22...0.0.23) (2021-09-16)


<a name="0.0.22"></a>
## [0.0.22](https://github.com/RasaHQ/rasactl/compare/0.0.21...0.0.22) (2021-09-16)


<a name="0.0.21"></a>
## [0.0.21](https://github.com/RasaHQ/rasactl/compare/0.0.20...0.0.21) (2021-09-08)


<a name="0.0.20"></a>
## [0.0.20](https://github.com/RasaHQ/rasactl/compare/0.0.19...0.0.20) (2021-09-08)


<a name="0.0.19"></a>
## [0.0.19](https://github.com/RasaHQ/rasactl/compare/0.0.18...0.0.19) (2021-09-07)


<a name="0.0.18"></a>
## [0.0.18](https://github.com/RasaHQ/rasactl/compare/0.0.17...0.0.18) (2021-09-07)


<a name="0.0.17"></a>
## [0.0.17](https://github.com/RasaHQ/rasactl/compare/0.0.16...0.0.17) (2021-09-06)


<a name="0.0.16"></a>
## [0.0.16](https://github.com/RasaHQ/rasactl/compare/0.0.15...0.0.16) (2021-09-06)


<a name="0.0.15"></a>
## [0.0.15](https://github.com/RasaHQ/rasactl/compare/0.0.14...0.0.15) (2021-09-06)


<a name="0.0.14"></a>
## [0.0.14](https://github.com/RasaHQ/rasactl/compare/0.0.13...0.0.14) (2021-09-03)


<a name="0.0.13"></a>
## [0.0.13](https://github.com/RasaHQ/rasactl/compare/0.0.12...0.0.13) (2021-09-02)


<a name="0.0.12"></a>
## [0.0.12](https://github.com/RasaHQ/rasactl/compare/0.0.11...0.0.12) (2021-09-02)


<a name="0.0.11"></a>
## [0.0.11](https://github.com/RasaHQ/rasactl/compare/0.0.10...0.0.11) (2021-08-25)


<a name="0.0.10"></a>
## [0.0.10](https://github.com/RasaHQ/rasactl/compare/0.0.9...0.0.10) (2021-08-25)


<a name="0.0.9"></a>
## [0.0.9](https://github.com/RasaHQ/rasactl/compare/0.0.8...0.0.9) (2021-08-23)


<a name="0.0.8"></a>
## [0.0.8](https://github.com/RasaHQ/rasactl/compare/0.0.7...0.0.8) (2021-08-23)


<a name="0.0.7"></a>
## [0.0.7](https://github.com/RasaHQ/rasactl/compare/0.0.6...0.0.7) (2021-08-20)


<a name="0.0.6"></a>
## [0.0.6](https://github.com/RasaHQ/rasactl/compare/0.0.5...0.0.6) (2021-08-18)


<a name="0.0.5"></a>
## [0.0.5](https://github.com/RasaHQ/rasactl/compare/0.0.4...0.0.5) (2021-08-18)


<a name="0.0.4"></a>
## [0.0.4](https://github.com/RasaHQ/rasactl/compare/0.0.3...0.0.4) (2021-08-17)


<a name="0.0.3"></a>
## 0.0.3 (2021-08-16)



cmd/flags.go
package cmd

import (
	"time"

	"github.com/spf13/cobra"

	"github.com/RasaHQ/rasactl/pkg/types"
)

func addStartUpgradeFlags(cmd *cobra.Command) {
	cmd.Flags().DurationVar(&helmConfiguration.Timeout, "wait-timeout", time.Minute*15, "time to wait for Rasa X to be ready")
	cmd.PersistentFlags().StringVar(&rasactlFlags.StartUpgrade.ValuesFile, "values-file", "", "absolute path to the values file")
}

func addStartFlags(cmd *cobra.Command) {
	cmd.Flags().StringVar(&helmConfiguration.ReleaseName, "rasa-x-release-name", "rasa-x", "a helm release name to manage")
	cmd.Flags().StringVar(&helmConfiguration.Version, "rasa-x-chart-version", types.HelmChartVersionRasaX, "a helm chart version to use")

	cmd.PersistentFlags().StringVar(&rasactlFlags.Start.ProjectPath, "project-path", "",
		"absolute path to the project directory mounted in kind")

	cmd.PersistentFlags().BoolVarP(&rasactlFlags.Start.Project, "project", "p", false,
		"use the current working directory as a project directory, the flag is ignored if --project-path is used")

	cmd.PersistentFlags().StringVar(&rasactlFlags.Start.RasaXPassword, "rasa-x-password", "rasaxlocal", "Rasa X password")
	cmd.PersistentFlags().BoolVar(&rasactlFlags.Start.RasaXPasswordStdin, "rasa-x-password-stdin", false, "read the Rasa X password from stdin")
	cmd.Flags().BoolVar(&rasactlFlags.Start.UseEdgeRelease, "rasa-x-edge-release", false, "use the latest edge release of Rasa X")
	cmd.Flags().BoolVar(&rasactlFlags.Start.Create, "create", false,
		"create a new deployment. If --project or --project-path is set, or there is no existing deployment,"+
			" the flag is not required to create a new deployment")

}

func addUpgradeFlags(cmd *cobra.Command) {
	cmd.Flags().BoolVar(&helmConfiguration.Atomic, "atomic", false, "if set, upgrade process rolls back changes made in case of failed upgrade")
	cmd.Flags().StringVar(&helmConfiguration.Version, "rasa-x-chart-version", "", "a helm chart version to use")
	cmd.Flags().BoolVar(&helmConfiguration.ReuseValues, "reuse-values", true,
		"when upgrading, reuse the last release's values and merge in any overrides")
}

func addDeleteFlags(cmd *cobra.Command) {
	cmd.PersistentFlags().BoolVar(&rasactlFlags.Delete.Force, "force", false, "if true, delete resources and ignore errors")
	cmd.PersistentFlags().BoolVar(&rasactlFlags.Delete.Prune, "prune", false, "if true, delete a namespace with a project")
}

func addStatusFlags(cmd *cobra.Command) {
	cmd.PersistentFlags().BoolVarP(&rasactlFlags.Status.Details, "details", "d", false,
		"show detailed information, such as running pods, helm chart status")
	cmd.PersistentFlags().StringVarP(&rasactlFlags.Status.Output, "output", "o", "table",
		"output format. One of: json|table")
}

func addAddFlags(cmd *cobra.Command) {
	cmd.Flags().StringVar(&helmConfiguration.ReleaseName, "rasa-x-release-name", "rasa-x", "a helm release name to manage")
}

func addConnectRasaFlags(cmd *cobra.Command) {
	cmd.Flags().IntVarP(&rasactlFlags.ConnectRasa.Port, "port", "p", 5005, "port to run the Rasa server at")
	cmd.Flags().BoolVar(&rasactlFlags.ConnectRasa.RunSeparateWorker, "run-separate-worker", false,
		"runs a separate Rasa server for the worker environment")
	cmd.Flags().StringSliceVar(&rasactlFlags.ConnectRasa.ExtraArgs, "extra-args", nil, "extra arguments for Rasa server")
}

func addAuthLoginFlags(cmd *cobra.Command) {
	cmd.PersistentFlags().BoolVar(&rasactlFlags.Auth.Login.PasswordStdin, "password-stdin", false, "read the password from stdin")
	cmd.PersistentFlags().StringVarP(&rasactlFlags.Auth.Login.Username, "username", "u", "", "username")
	cmd.PersistentFlags().StringVarP(&rasactlFlags.Auth.Login.Password, "password", "p", "", "password")
}

func configFlags(cmd *cobra.Command) {
	cmd.PersistentFlags().BoolVar(&rasactlFlags.Config.CreateFile, "create-file", false, "create the configuration file if it doesn't exist")
}

func enterpriseActivateFlags(cmd *cobra.Command) {
	cmd.PersistentFlags().BoolVar(&rasactlFlags.Enterprise.Activate.LicenseStdin, "license-stdin", false, "read an Enterprise license from stdin")
	cmd.PersistentFlags().StringVarP(&rasactlFlags.Enterprise.Activate.License, "license", "l", "", "an Enterprise license")
}

func logsFlags(cmd *cobra.Command) {
	cmd.PersistentFlags().BoolVarP(&rasactlFlags.Logs.Follow, "follow", "f", false, "specify if the logs should be streamed")
	cmd.PersistentFlags().BoolVarP(&rasactlFlags.Logs.Previous, "previous", "p", false, "print the logs for the previous instance of "+
		"the container in a pod if it exists")
	cmd.PersistentFlags().Int64Var(&rasactlFlags.Logs.TailLines, "tail", -1, "lines of recent log file to display. Defaults to -1 showing all log lines")
	cmd.PersistentFlags().StringVarP(&rasactlFlags.Logs.Container, "container", "c", "", "a container name")
}
